File: DynamoDB-SDK-Examples/python/WorkingWithTables/enable_auto-scaling.py
# In order for this script to work, the calling account must have permissions to create policies and roles in IAM
# and the DynamoDB table must already be in provisioned capacity mode.

import boto3
import json
aas_client = boto3.client('application-autoscaling')
iam_client = boto3.client('iam')

table_name = "YourTableName" # Add the name of the DynamoDB table you want to add auto-scaling to between the double quotes.
min_capacity = 1 # The minimum capacity for the auto-scaling policy
max_capacity = 100 # The maximum capacity for the auto-scaling policy
read_target = 50 # The target percentage utilization for read capacity
write_target = 50 # The target percentage utilization for write capacity
cooldown_duration_sec = 150 # How long in seconds

role_name = '{}TableScalingRole'.format(table_name)
policy_name = '{}TableScalingPolicy'.format(table_name)

assume_role_policy_document = {"Version": "2012-10-17", "Statement": [{"Effect": "Allow","Principal": {"Service": [ "ec2.amazonaws.com" ]},"Action": ["sts:AssumeRole"]}]}
policy_document = {"Version": "2012-10-17", "Statement":[{"Effect": "Allow", "Action": ["dynamodb:DescribeTable", "dynamodb:UpdateTable", "cloudwatch:PutMetricAlarm", "cloudwatch:DescribeAlarms", "cloudwatch:GetMetricStatistics", "cloudwatch:SetAlarmState", "cloudwatch:DeleteAlarms"], "Resource": "*"}]}

# Create the role necessary for auto-scaling
create_role_response = iam_client.create_role(
    Path='/',
    RoleName=role_name,
    AssumeRolePolicyDocument=json.dumps(assume_role_policy_document),
    Description='Table Scaling Role for {}'.format(table_name),
    MaxSessionDuration=3600
)

# Create the policy needed by the role
create_policy_response = iam_client.create_policy(
    PolicyName=policy_name,
    Path='/',
    PolicyDocument=json.dumps(policy_document)
)

role_arn = create_role_response['Role']['Arn']
policy_arn = create_policy_response['Policy']['Arn']

# Attach the policy to the role so it can be used.
response = iam_client.attach_role_policy(
    RoleName=role_name,
    PolicyArn=policy_arn
)

# Register the RCU targets for the table
response = aas_client.register_scalable_target(
    ServiceNamespace='dynamodb',
    ResourceId='table/{}'.format(table_name),
    ScalableDimension='dynamodb:table:ReadCapacityUnits',
    MinCapacity=min_capacity,
    MaxCapacity=max_capacity,
    RoleARN=role_arn
)

# Register the WCU targets for the table
response = aas_client.register_scalable_target(
    ServiceNamespace='dynamodb',
    ResourceId='table/{}'.format(table_name),
    ScalableDimension='dynamodb:table:WriteCapacityUnits',
    MinCapacity=min_capacity,
    MaxCapacity=max_capacity,
    RoleARN=role_arn
)

# Attach the Read scaling policy to the table.
response = aas_client.put_scaling_policy(
    PolicyName='{}ScalingPolicy'.format(table_name),
    ServiceNamespace='dynamodb',
    ResourceId='table/{}'.format(table_name),
    ScalableDimension='dynamodb:table:ReadCapacityUnits',
    PolicyType='TargetTrackingScaling',
    TargetTrackingScalingPolicyConfiguration={
        'TargetValue': read_target,
        'PredefinedMetricSpecification': {'PredefinedMetricType': 'DynamoDBReadCapacityUtilization'},
        'ScaleOutCooldown': cooldown_duration_sec,
        'ScaleInCooldown': cooldown_duration_sec,
        'DisableScaleIn': True
    }
)

# Attach the Write scaling policy to the table.
response = aas_client.put_scaling_policy(
    PolicyName='{}ScalingPolicy'.format(table_name),
    ServiceNamespace='dynamodb',
    ResourceId='table/{}'.format(table_name),
    ScalableDimension='dynamodb:table:WriteCapacityUnits',
    PolicyType='TargetTrackingScaling',
    TargetTrackingScalingPolicyConfiguration={
        'TargetValue': write_target,
        'PredefinedMetricSpecification': {'PredefinedMetricType': 'DynamoDBWriteCapacityUtilization'},
        'ScaleOutCooldown': cooldown_duration_sec,
        'ScaleInCooldown': cooldown_duration_sec,
        'DisableScaleIn': True
    }
)

File: DynamoDB-SDK-Examples/node.js/WorkingWithTables/enable_auto_scaling.js
const AWS = require("aws-sdk");

const iam = new AWS.IAM({ apiVersion: "2010-05-08", region: "us-west-2", logger: console });
const applicationAutoscaling = new AWS.ApplicationAutoScaling({
  apiVersion: "2016-02-06",
  region: "us-west-2",
  logger: console,
});

const tableName = "Music";
const roleName = `${tableName}TableScalingRole`;
const policyName = `${tableName}TableScalingPolicy`;

const minCapacity = 1; // The minimum capacity for the auto-scaling policy
const maxCapacity = 100; // The maximum capacity for the auto-scaling policy
const readTarget = 50; // The target percentage utilization for read capacity
const writeTarget = 50; // The target percentage utilization for write capacity
const cooldownDurationSec = 150; // How long in seconds

const assumeRolePolicyDocument = {
  Version: "2012-10-17",
  Statement: [
    {
      Effect: "Allow",
      Principal: {
        Service: ["ec2.amazonaws.com"]
      },
      Action: ["sts:AssumeRole"] },
  ],
};

const policyDocument = {
  Version: "2012-10-17",
  Statement: [
    {
      Effect: "Allow",
      Action: [
        "dynamodb:DescribeTable",
        "dynamodb:UpdateTable",
        "cloudwatch:PutMetricAlarm",
        "cloudwatch:DescribeAlarms",
        "cloudwatch:GetMetricStatistics",
        "cloudwatch:SetAlarmState",
        "cloudwatch:DeleteAlarms",
      ],
      Resource: "*",
    },
  ],
};

const enableAutoScaling = async () => {
  // Create the role necessary for auto-scaling
  const {
    Role: { Arn: roleArn },
  } = await iam
    .createRole({
      AssumeRolePolicyDocument: JSON.stringify(assumeRolePolicyDocument),
      RoleName: roleName,
      Description: `Table scaling role for ${tableName}`,
      MaxSessionDuration: 3600,
      Path: "/",
    })
    .promise();

  // Create the policy needed by the role
  const {
    Policy: { Arn: policyArn },
  } = await iam
    .createPolicy({
      PolicyDocument: JSON.stringify(policyDocument),
      PolicyName: policyName,
      Path: "/",
    })
    .promise();

  // Attach the policy to the role so it can be used.
  let response = await iam
    .attachRolePolicy({
      RoleName: roleName,
      PolicyArn: policyArn,
    })
    .promise();

  // Register the RCU targets for the table
  response = await applicationAutoscaling
    .registerScalableTarget({
      ServiceNamespace: "dynamodb",
      ResourceId: `table/${tableName}`,
      ScalableDimension: "dynamodb:table:ReadCapacityUnits",
      MinCapacity: minCapacity,
      MaxCapacity: maxCapacity,
      RoleARN: "arn:aws:iam::544500146257:role/MusicTableScalingRole",
    })
    .promise();

  // Register the RCU targets for the table
  response = await applicationAutoscaling
    .registerScalableTarget({
      ServiceNamespace: "dynamodb",
      ResourceId: `table/${tableName}`,
      ScalableDimension: "dynamodb:table:WriteCapacityUnits",
      MinCapacity: minCapacity,
      MaxCapacity: maxCapacity,
      RoleARN: "arn:aws:iam::544500146257:role/MusicTableScalingRole",
    })
    .promise();

  // Attach the Read scaling policy to the table.
  response = await applicationAutoscaling
    .putScalingPolicy({
      PolicyName: `${tableName}ScalingPolicy`,
      ServiceNamespace: "dynamodb",
      ResourceId: `table/${tableName}`,
      ScalableDimension: "dynamodb:table:ReadCapacityUnits",
      PolicyType: "TargetTrackingScaling",
      TargetTrackingScalingPolicyConfiguration: {
        TargetValue: readTarget,
        PredefinedMetricSpecification: { PredefinedMetricType: "DynamoDBReadCapacityUtilization" },
        ScaleOutCooldown: cooldownDurationSec,
        ScaleInCooldown: cooldownDurationSec,
        DisableScaleIn: true,
      },
    })
    .promise();

  // Attach the Write scaling policy to the table.
  response = await applicationAutoscaling
    .putScalingPolicy({
      PolicyName: `${tableName}ScalingPolicy`,
      ServiceNamespace: "dynamodb",
      ResourceId: `table/${tableName}`,
      ScalableDimension: "dynamodb:table:WriteCapacityUnits",
      PolicyType: "TargetTrackingScaling",
      TargetTrackingScalingPolicyConfiguration: {
        TargetValue: writeTarget,
        PredefinedMetricSpecification: { PredefinedMetricType: "DynamoDBWriteCapacityUtilization" },
        ScaleOutCooldown: cooldownDurationSec,
        ScaleInCooldown: cooldownDurationSec,
        DisableScaleIn: true,
      },
    })
    .promise();

  console.log("Autoscaling has been enabled");
};

enableAutoScaling().catch((error) => console.error(JSON.stringify(error, null, 2)));
